File: weather-server-python/weather.py
from typing import Any

import httpx2
from mcp.server import MCPServer
from pydantic import BaseModel, Field, RootModel

# Initialize the MCP server
mcp = MCPServer("weather")

# Constants
NWS_API_BASE = "https://api.weather.gov"
USER_AGENT = "weather-app/1.0"


class Alert(BaseModel):
    """One active weather alert."""

    event: str = Field(description="The kind of weather event")
    area: str = Field(description="The area the alert covers")
    severity: str = Field(description="How severe the event is")
    description: str = Field(description="What is happening")
    instructions: str = Field(description="What people in the area should do")


class Alerts(RootModel[list[Alert]]):
    """The output schema of get_alerts: a top-level array, not an object.

    A plain `-> list[Alert]` annotation would be wrapped as `{"result": [...]}`.
    A RootModel is a BaseModel, so it is taken as the schema as written, and a
    RootModel over a list serializes as the bare list. Revision 2026-07-28 is
    the first to allow a non-object root.
    """


class Period(BaseModel):
    """One period of a forecast."""

    name: str
    temperature: int
    temperature_unit: str
    wind_speed: str
    wind_direction: str
    detailed_forecast: str


class Forecast(BaseModel):
    """The output schema of get_forecast: the object case, for contrast."""

    latitude: float = Field(description="Latitude the forecast is for")
    longitude: float = Field(description="Longitude the forecast is for")
    periods: list[Period] = Field(description="The forecast periods, soonest first")


async def make_nws_request(url: str) -> dict[str, Any] | None:
    """Make a request to the NWS API with proper error handling."""
    headers = {"User-Agent": USER_AGENT, "Accept": "application/geo+json"}
    async with httpx2.AsyncClient() as client:
        try:
            response = await client.get(url, headers=headers, timeout=30.0)
            response.raise_for_status()
            return response.json()
        except Exception:
            return None


@mcp.tool()
async def get_alerts(state: str) -> Alerts:
    """Get weather alerts for a US state.

    Args:
        state: Two-letter US state code (e.g. CA, NY)
    """
    url = f"{NWS_API_BASE}/alerts/active/area/{state.upper()}"
    data = await make_nws_request(url)

    if not data or "features" not in data:
        raise ValueError(f"Unable to fetch alerts for {state.upper()}.")

    # An empty result is an empty array, not an error.
    #
    # `or` rather than a `get` default: NWS sends these fields as explicit
    # nulls rather than omitting them, and a default only covers a missing key.
    return Alerts(
        [
            Alert(
                event=props.get("event") or "Unknown",
                area=props.get("areaDesc") or "Unknown",
                severity=props.get("severity") or "Unknown",
                description=props.get("description") or "No description available",
                instructions=props.get("instruction") or "No specific instructions provided",
            )
            for props in (feature["properties"] for feature in data["features"])
        ]
    )


@mcp.tool()
async def get_forecast(latitude: float, longitude: float) -> Forecast:
    """Get weather forecast for a location.

    Args:
        latitude: Latitude of the location
        longitude: Longitude of the location
    """
    # First get the forecast grid endpoint
    points_url = f"{NWS_API_BASE}/points/{latitude},{longitude}"
    points_data = await make_nws_request(points_url)

    if not points_data:
        raise ValueError("Unable to fetch forecast data for this location.")

    # Get the forecast URL from the points response
    forecast_url = points_data["properties"]["forecast"]
    forecast_data = await make_nws_request(forecast_url)

    if not forecast_data:
        raise ValueError("Unable to fetch detailed forecast.")

    periods = forecast_data["properties"]["periods"][:5]  # Only show next 5 periods
    if not periods:
        raise ValueError("No forecast periods available.")

    return Forecast(
        latitude=latitude,
        longitude=longitude,
        periods=[
            Period(
                name=period["name"],
                temperature=period["temperature"],
                temperature_unit=period["temperatureUnit"],
                wind_speed=period["windSpeed"],
                wind_direction=period["windDirection"],
                detailed_forecast=period["detailedForecast"],
            )
            for period in periods
        ],
    )


def main():
    # Initialize and run the server
    mcp.run(transport="stdio")


if __name__ == "__main__":
    main()

File: weather-server-ruby/weather.rb
# frozen_string_literal: true

require "json"
require "mcp"
require "net/http"
require "uri"

NWS_API_BASE = "https://api.weather.gov"
USER_AGENT = "weather-app/1.0"

module HelperMethods
  def make_nws_request(url)
    uri = URI(url)
    request = Net::HTTP::Get.new(uri)
    request["User-Agent"] = USER_AGENT
    request["Accept"] = "application/geo+json"

    response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true, read_timeout: 30) do |http|
      http.request(request)
    end

    raise "HTTP #{response.code}: #{response.message}" unless response.is_a?(Net::HTTPSuccess)

    JSON.parse(response.body)
  end

  def format_alert(alert)
    <<~ALERT
      Event: #{alert[:event]}
      Area: #{alert[:area]}
      Severity: #{alert[:severity]}
      Description: #{alert[:description]}
      Instructions: #{alert[:instructions]}
    ALERT
  end

  # A tool that declares an output schema MUST return conforming structured
  # content, so a failure path has to be an error result rather than a bare
  # text one. Error results are exempt from output schema validation.
  def error_response(text)
    MCP::Tool::Response.new([{ type: "text", text: text }], error: true)
  end
end

class GetAlerts < MCP::Tool
  extend HelperMethods

  tool_name "get_alerts"
  description "Get weather alerts for a US state"
  input_schema(
    properties: {
      state: {
        type: "string",
        description: "Two-letter US state code (e.g. CA, NY)"
      }
    },
    required: ["state"]
  )

  # A top-level array, not an object wrapping one. Declaring `type` at the root
  # suppresses the object default the SDK would otherwise apply, and revision
  # 2026-07-28 is the first to allow a non-object root.
  output_schema(
    type: "array",
    items: {
      type: "object",
      properties: {
        event: { type: "string", description: "The kind of weather event" },
        area: { type: "string", description: "The area the alert covers" },
        severity: { type: "string", description: "How severe the event is" },
        description: { type: "string", description: "What is happening" },
        instructions: { type: "string", description: "What people in the area should do" }
      },
      required: ["event", "area", "severity", "description", "instructions"]
    }
  )

  def self.call(state:, server_context: nil)
    data = begin
      make_nws_request("#{NWS_API_BASE}/alerts/active/area/#{state.upcase}")
    rescue => e
      return error_response("Unable to fetch alerts for #{state.upcase}: #{e.message}")
    end

    unless data.is_a?(Hash) && data["features"]
      return error_response("Unable to fetch alerts for #{state.upcase}.")
    end

    # An empty result is an empty array, not an error.
    alerts = data["features"].map do |feature|
      properties = feature["properties"] || {}
      {
        event: properties["event"] || "Unknown",
        area: properties["areaDesc"] || "Unknown",
        severity: properties["severity"] || "Unknown",
        description: properties["description"] || "No description available",
        instructions: properties["instruction"] || "No specific instructions provided"
      }
    end

    text = if alerts.empty?
      "No active alerts for this state."
    else
      alerts.map { |alert| format_alert(alert) }.join("\n---\n")
    end

    MCP::Tool::Response.new(
      [{ type: "text", text: text }],
      structured_content: alerts
    )
  end
end

class GetForecast < MCP::Tool
  extend HelperMethods

  tool_name "get_forecast"
  description "Get weather forecast for a location"
  input_schema(
    properties: {
      latitude: {
        type: "number",
        description: "Latitude of the location"
      },
      longitude: {
        type: "number",
        description: "Longitude of the location"
      }
    },
    required: ["latitude", "longitude"]
  )

  # The object case, for contrast with get_alerts' array root.
  output_schema(
    type: "object",
    properties: {
      latitude: { type: "number", description: "Latitude the forecast is for" },
      longitude: { type: "number", description: "Longitude the forecast is for" },
      periods: {
        type: "array",
        description: "The forecast periods, soonest first",
        items: {
          type: "object",
          properties: {
            name: { type: "string" },
            temperature: { type: "integer" },
            temperature_unit: { type: "string" },
            wind_speed: { type: "string" },
            wind_direction: { type: "string" },
            detailed_forecast: { type: "string" }
          },
          required: ["name", "temperature", "temperature_unit", "wind_speed", "wind_direction", "detailed_forecast"]
        }
      }
    },
    required: ["latitude", "longitude", "periods"]
  )

  def self.call(latitude:, longitude:, server_context: nil)
    forecast_data = begin
      # First get the forecast grid endpoint.
      points_data = make_nws_request("#{NWS_API_BASE}/points/#{latitude},#{longitude}")
      forecast_url = points_data.dig("properties", "forecast")
      return error_response("Unable to fetch forecast data for this location.") unless forecast_url

      make_nws_request(forecast_url)
    rescue => e
      return error_response("Unable to fetch forecast data for this location: #{e.message}")
    end

    # Only show the next 5 periods.
    periods = (forecast_data.dig("properties", "periods") || []).first(5)
    return error_response("No forecast periods available.") if periods.empty?

    forecast = {
      latitude: latitude,
      longitude: longitude,
      periods: periods.map do |period|
        {
          name: period["name"],
          temperature: period["temperature"],
          temperature_unit: period["temperatureUnit"],
          wind_speed: period["windSpeed"],
          wind_direction: period["windDirection"],
          detailed_forecast: period["detailedForecast"]
        }
      end
    }

    text = forecast[:periods].map do |period|
      <<~FORECAST
        #{period[:name]}:
        Temperature: #{period[:temperature]}°#{period[:temperature_unit]}
        Wind: #{period[:wind_speed]} #{period[:wind_direction]}
        Forecast: #{period[:detailed_forecast]}
      FORECAST
    end.join("\n---\n")

    MCP::Tool::Response.new(
      [{ type: "text", text: text }],
      structured_content: forecast
    )
  end
end

server = MCP::Server.new(
  name: "weather",
  version: "1.0.0",
  tools: [GetAlerts, GetForecast],
  # The gem otherwise advertises prompts, resources and logging as well. This
  # server has only tools, and logging is deprecated as of 2026-07-28
  # (SEP-2577). No listChanged: change notifications are delivered over
  # `subscriptions/listen`, which this server does not serve.
  capabilities: { tools: {} }
)

transport = MCP::Server::Transports::StdioTransport.new(server)
transport.open
